CODE_STYLE.md

```diff
@@ -1,151 +1,392 @@
+
+Specifically, if all items except one are private, always put the non-private item on top.
 ```rust
-impl ErrorKind {
-    fn print(&self) {
-        ErrorKind::Io => println!("Io"),
-        ErrorKind::Network => println!("Network"),
-        ErrorKind::Json => println!("Json"),
+// GOOD
+pub(crate) fn frobnicate() {
+    Helper::act()
+}
+
+#[derive(Default)]
+struct Helper { stuff: i32 }
+
+impl Helper {
+    fn act(&self) {
+
+    }
+}
+
+// BAD
+#[derive(Default)]
+struct Helper { stuff: i32 }
+
+pub(crate) fn frobnicate() {
+    Helper::act()
+}
+
+impl Helper {
+    fn act(&self) {
+
     }
 }
 ```
 
-Good:
+If there's a mixture of private and public items, put public items first.
+
+Put structs and enums first, functions and impls last. Order type declarations in top-down manner.
+
 ```rust
-impl ErrorKind {
-    fn print(&self) {
-        Self::Io => println!("Io"),
-        Self::Network => println!("Network"),
-        Self::Json => println!("Json"),
-    }
+// GOOD
+struct Parent {
+    children: Vec<Child>
+}
+
+struct Child;
+
+impl Parent {
+}
+
+impl Child {
+}
+
+// BAD
+struct Child;
+
+impl Child {
+}
+
+struct Parent {
+    children: Vec<Child>
+}
+
+impl Parent {
 }
 ```
 
-<details>
-    <summary>More examples</summary>
-    
-Bad:
-    
+**Rationale:** easier to get the sense of the API by visually scanning the file. If function bodies are folded in the editor, the source code should read as documentation for the public API.
+
+## Early Returns
+
+Do use early returns
+
 ```rust
-impl<'a> AnswerCallbackQuery<'a> {
-    pub(crate) fn new<C>(bot: &'a Bot, callback_query_id: C) -> AnswerCallbackQuery<'a>
-    where
-C: Into<String>, { ... }
+// GOOD
+fn foo() -> Option<Bar> {
+    if !condition() {
+        return None;
+    }
+
+    Some(...)
+}
+
+// BAD
+fn foo() -> Option<Bar> {
+    if condition() {
+        Some(...)
+    } else {
+        None
+    }
+}
 ```
 
-Good:
-    
+**Rationale:** reduce cognitive stack usage.
+
+## If-let
+
+Avoid if let ... { } else { } construct, use match instead.
+
 ```rust
-impl<'a> AnswerCallbackQuery<'a> {
-    pub(crate) fn new<C>(bot: &'a Bot, callback_query_id: C) -> Self
-    where
-C: Into<String>, { ... }
+// GOOD
+match ctx.expected_type.as_ref() {
+    Some(expected_type) => completion_ty == expected_type && !expected_type.is_unit(),
+    None => false,
+}
+
+// BAD
+if let Some(expected_type) = ctx.expected_type.as_ref() {
+    completion_ty == expected_type && !expected_type.is_unit()
+} else {
+    false
+}
 ```
-</details>
 
-## Naming
- 1. Avoid unnecessary duplication (`Message::message_id` -> `Message::id` using `#[serde(rename = "message_id")]`).
- 2. Use a generic parameter name `S` for streams, `Fut` for futures, `F` for functions (where possible).
+**Rationale:** `match is almost always more compact. The `else` branch can get a more precise pattern: `None` or `Err(_)` instead of `_`.
```

Comment:
```suggestion
**Rationale:**
- `match` is almost always more compact
- The `else` branch can get a more precise pattern: `None` or `Err(_)` instead of `_`
```